sonar-plugin-api/src/main/java/org/sonar/api/server/rule/internal/DefaultNewRule.java

```diff
@@ -134,13 +136,48 @@ public DefaultNewRule setType(RuleType t) {
 
   @Override
   public RulesDefinition.NewRule addDescriptionSection(RuleDescriptionSection ruleDescriptionSection) {
-    checkArgument(isSectionKeyUnique(ruleDescriptionSection.getKey()), "A section with key %s already exists", ruleDescriptionSection.getKey());
+    assertRuleDescriptionSectionIsValid(ruleDescriptionSection);
     ruleDescriptionSections.add(ruleDescriptionSection);
     return this;
   }
 
+  private void assertRuleDescriptionSectionIsValid(RuleDescriptionSection ruleDescriptionSection) {
+    ruleDescriptionSection.getContext().ifPresent(context -> assertContextAwareRuleDescriptionIsValid(ruleDescriptionSection, context));
+    if (ruleDescriptionSection.getContext().isEmpty()) {
+      checkArgument(isSectionKeyUnique(ruleDescriptionSection.getKey()), "A section with key %s already exists", ruleDescriptionSection.getKey());
+    }
+  }
+
+  private void assertContextAwareRuleDescriptionIsValid(RuleDescriptionSection ruleDescriptionSection, Context context) {
+    String sectionKey = ruleDescriptionSection.getKey();
+    String contextKey = context.getKey();
+    checkArgument(isContextSetForAllRuleDescriptionSectionWithKey(sectionKey),
+      "Section with key %s already added without context information. "
```

Comment:
A change of the string in here would result in changing it in the test too since we test the message.
We could put this string in a protected variable and reuse it in the test to avoid duplication.

Reply:
Yes indeed, fixed :)

Reply:
But then if we re-use the constant in the test, then we are not testing the message anymore: someone can later put the wrong message in the constant and the test will still be passing.

It is double edged sword.